KX_GameObject kwargs

TryExceptElse opened the pull request:
All KX_GameObject python methods with multiple or optional args can now accept kwargs
test blend: http://pasteall.org/blend/index.php?id=47840

panzergame commented on source/gameengine/Ketsji/KX_GameObject.cpp:
```diff
@@ -3759,7 +3759,11 @@ KX_PYMETHODDEF_DOC(KX_GameObject, rayCastTo,
+	if (!PyArg_ParseTupleAndKeywords(
+	        args, kwds, "O|fs:rayCastTo",  const_cast<char**>(kwlist),
+	        &pyarg, &dist, &propName
+	    )) {
```
The code style we are using try to put all in one line if possible, if not you have to break and the `{` is not anymore in the same line than the condition but in a new line. Here i think you can put everything in the same line.

panzergame commented on source/gameengine/Ketsji/KX_GameObject.cpp:
```diff
@@ -4233,8 +4233,14 @@ KX_PYMETHODDEF_DOC(KX_GameObject, addDebugProperty,
 
-	if (!PyArg_ParseTuple(args,"s|i:debugProperty", &name , &visible))
+	static const char *kwlist[] = {"name", "debug", nullptr};
+	if (!PyArg_ParseTupleAndKeywords(
```
I don't know if this function request keyword, `name` must be provided and after that there's only one value.

panzergame commented on source/gameengine/Ketsji/KX_GameObject.cpp:
```diff
@@ -3252,12 +3252,16 @@ int KX_GameObject::pyattr_set_lodManager(PyObjectPlus *self_v, const KX_PYATTRIB
 	PyObject *pyvect;
 
-	if (PyArg_ParseTuple(args, "O|i:applyForce", &pyvect, &local)) {
+    static const char *kwlist[] = {"force", "local", nullptr};
```
As for `addDebugProperty` i'm not sure of the interest of making the second argument a keyword. If the doc say the opposite you can change the doc like:
`applyBlah(vector[, local])`

panzergame commented on source/gameengine/Ketsji/KX_GameObject.cpp:
```diff
@@ -3535,11 +3535,15 @@ PyObject *KX_GameObject::PyApplyImpulse(PyObject *args, PyObject *kwds)
 	int freeConstraints = false;
 
-	if (!PyArg_ParseTuple(args, "|i:suspendPhysics", &freeConstraints)) {
+    static const char *kwlist[] = {"freeConstraints", nullptr};
```
Same than `apply**`.

panzergame commented on source/gameengine/Ketsji/KX_GameObject.cpp:
```diff
@@ -3561,12 +3561,17 @@ PyObject *KX_GameObject::PyRestorePhysics()
 	bool ghost = false;
 
-	if (!PyArg_ParseTuple(args, "|b", &ghost))
+    static const char *kwlist[] = {"ghost", nullptr};
```
Same than `apply**`.

panzergame commented on source/gameengine/Ketsji/KX_GameObject.cpp:
```diff
@@ -1993,7 +1993,11 @@ PyObject *KX_GameObject::PyReinstancePhysicsMesh(PyObject *args)
 	PyObject *mesh_py= nullptr;
 
-	if (!PyArg_ParseTuple(args,"|OOi:reinstancePhysicsMesh",&gameobj_py, &mesh_py, &dupli) ||
+    static const char *kwlist[] = {"gameObject", "meshObject", "dupli", nullptr};
```
This one is good :).

panzergame commented on source/gameengine/Ketsji/KX_GameObject.cpp:
```diff
@@ -4125,8 +4125,14 @@ KX_PYMETHODDEF_DOC(KX_GameObject, stopAction,
 	short layer = 0;
 
-	if (!PyArg_ParseTuple(args, "|h:stopAction", &layer))
+    static const char *kwlist[] = {"layer", nullptr};
```
Same as `apply**`

panzergame commented on source/gameengine/Ketsji/KX_GameObject.cpp:
```diff
@@ -4170,8 +4188,14 @@ KX_PYMETHODDEF_DOC(KX_GameObject, setActionFrame,
 	float frame;
 
-	if (!PyArg_ParseTuple(args, "f|h:setActionFrame", &frame, &layer))
+    static const char *kwlist[] = {"frame", "layer", nullptr};
```
Same as `apply**`

TryExceptElse commented:
removed kwarg parsing from methods that only take 1 or 2 args and updated documentation to reflect which methods take kwargs, and which only accept positional args.

TryExceptElse commented:
[new test .blend](http://pasteall.org/blend/index.php?id=47843)

youle31 commented on doc/python_api/rst/bge_types/bge.types.KX_GameObject.rst:
```diff
@@ -595,7 +599,7 @@ base class --- :class:`SCA_IObject`
       :rtype: 3d vector.
 
-   .. method:: applyMovement(movement, local=False)
+   .. method:: applyMovement(movement[, local])
```
hello, it seems you forgot to indicate the default local value here and for applyRotation too.

youle31 commented:
For me appart from the remark I noted (in the rst file), it sounds good. I have another small remak but it isn't so important to me and in KX_GameObject, most of the methods have the code style you choosed but usually when we are not in a switch: case or in a function definition, we put the first brackets "{" on the same line than the if. But as other methods in this file don't follow this rule, it's just a remark and ont a request to change that. I didn't tested, but this sounds good to me.

youle31 commented:
For me it's ok. Thanks for your work TryExceptElse

panzergame commented:
@TryExceptElse : I fixed some doc and source issue in branch ge_kwargs, could you check it please ?
